Tests/Unit/NetworkServer/CupsProtocolMessageProcessorTests.cs

```diff
@@ -110,21 +118,167 @@ public async Task HandleUpdateInfoAsync_Fails_WithInvalidInput(string input, Typ
             Assert.Contains(this.logger.Invocations, i => i.Arguments.Any(a => a.GetType() == exceptionType));
         }
 
+        [Fact]
+        public async Task HandleUpdateInfoAsync_Invokes_FetchFirmware_WhenUpdateAvailable()
+        {
+            // setup
+            using var memoryStream = new MemoryStream();
+            var strictifiedInput = JsonUtil.Strictify(CupsRequestJson);
+            var (httpContext, httpRequest, _) = SetupHttpContextWithRequest(strictifiedInput, memoryStream);
+            _ = httpRequest.Setup(r => r.ContentLength).Returns(Encoding.UTF8.GetByteCount(strictifiedInput));
+
+            // setting up the twin in such a way that there is a fw update but no matching checksum
+            var cupsTwinInfo = new CupsTwinInfo(new Uri(CupsUri),
+                                                new Uri(TcUri),
+                                                CredentialsChecksum,
+                                                CredentialsChecksum,
+                                                string.Empty,
+                                                string.Empty,
+                                                "anotherVersion",
+                                                KeyChecksum,
+                                                "ABCD");
+            _ = this.basicsStationConfigurationService.Setup(m => m.GetCupsConfigAsync(It.IsAny<StationEui>(), It.IsAny<CancellationToken>()))
+                                                      .Returns(Task.FromResult(cupsTwinInfo));
+
+            var firmwareBytes = new byte[] { 1, 2, 3 };
+            using var fwStream = new MemoryStream(firmwareBytes);
+            using var streamContent = new StreamContent(fwStream);
+            _ = this.deviceAPIServiceBase.Setup(m => m.FetchStationFirmwareAsync(It.IsAny<StationEui>(), It.IsAny<CancellationToken>()))
+                                         .ReturnsAsync(streamContent);
+            // act
+            await this.processor.HandleUpdateInfoAsync(httpContext.Object, default);
+
+            // assert
+            this.deviceAPIServiceBase.Verify(m => m.FetchStationFirmwareAsync(StationEui.Parse(StationEuiString), It.IsAny<CancellationToken>()), Times.Once);
+            var cupsBytes = new byte[memoryStream.Position];
+            memoryStream.Seek(0, SeekOrigin.Begin);
+            await memoryStream.ReadAsync(cupsBytes);
+            Assert.Equal(firmwareBytes, cupsBytes[^3..]);
```

Comment:
You can just grab the array of the memory stream and boil this down to just one line:

```suggestion
            Assert.Equal(firmwareBytes, memoryStream.ToArray().TakeLast(firmwareBytes.Length));
```

However, this does not test the beginning of the response, that the serialized `CupsUpdateInfoResponseHeader`? My fear is that if someone introduces a regression in `CupsProtocolMessageProcessor` then this will only test the firmware bytes were delivered but not that the LBS will accept the entire response in case the header is corrupt.

Reply:
Fixed and added additional assertions